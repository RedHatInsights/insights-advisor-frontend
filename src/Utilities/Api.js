import axios from 'axios';

export default {
    get(url, headers = {}, params = {}) {
        return axios.get(url, {
            headers,
            params
        });
    },
    put(url, data = {}, headers = {}) {
        return axios.put(url, data, {
            headers
        });
    },
    post(url, data = {}, headers = {}) {
        return axios.post(url, data, {
            headers
        });
    }
};
