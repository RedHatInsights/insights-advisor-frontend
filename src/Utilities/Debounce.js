import { useEffect, useState } from 'react';

const Debounce = (value, delay) => {
    const [debouncedValue, setDebouncedValue] = useState(value);
    useEffect(() => {
        const handler = setTimeout(() => {setDebouncedValue(value);}, delay);
        return () => {
            clearTimeout(handler);
        };
    }, [delay, value]);

    return debouncedValue;
};

export default Debounce;
