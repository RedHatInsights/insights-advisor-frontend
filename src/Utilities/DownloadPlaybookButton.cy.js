import React from 'react';
import DownloadPlaybookButton from './DownloadPlaybookButton';
import { REMEDIATIONS_BASE_URL } from '../AppConstants';
import { IntlProvider } from 'react-intl';
import { Provider } from 'react-redux';
import { initStore } from '../Store';

const mockRules = [
  {
    rule_id: 'rule1',
    resolution_set: [{ has_playbook: true }],
    reboot_required: false,
  },
];
const mockSystems = ['system1'];

const mountComponent = (
  isDisabled = false,
  rules = mockRules,
  systems = mockSystems,
) => {
  cy.mount(
    <IntlProvider locale={navigator.language.slice(0, 2)}>
      <Provider store={initStore()}>
        <DownloadPlaybookButton
          isDisabled={isDisabled}
          rules={rules}
          systems={systems}
        />
      </Provider>
    </IntlProvider>,
  );
};

describe('DownloadPlaybookButton Component', () => {
  it('renders the button and triggers download when clicked', () => {
    mountComponent();

    cy.intercept('POST', `${REMEDIATIONS_BASE_URL}/playbook`, {
      statusCode: 200,
      body: 'mock-playbook-content',
    }).as('postPlaybook');

    cy.get('button').contains('Download playbook').click();
    cy.wait('@postPlaybook')
      .its('request.body')
      .should('deep.equal', {
        auto_reboot: false,
        issues: [{ id: 'advisor:rule1', systems: ['system1'] }],
      });
  });

  it('disables the button when isDisabled is true', () => {
    mountComponent({ isDisabled: true, rules: [{}], systems: [] });
    cy.get('button').should('be.disabled');
  });
});
