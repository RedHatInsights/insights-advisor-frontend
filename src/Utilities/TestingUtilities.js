import React, { createContext } from 'react';
import PropTypes from 'prop-types';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import { MemoryRouter, Routes, Route } from 'react-router-dom';
import { IntlProvider } from 'react-intl';

export const ComponentWithContext = ({
  Component,
  componentProps,
  renderOptions = {},
  Context = createContext({}),
  contextValue = {},
}) => {
  const mockStore = configureStore();

  return (
    <IntlProvider locale="en">
      <Provider store={renderOptions?.store || mockStore()}>
        <MemoryRouter initialEntries={renderOptions?.initialEntries || ['/']}>
          <Context.Provider value={contextValue}>
            {renderOptions?.componentPath ? (
              <Routes>
                <Route>
                  <Component {...componentProps} />
                </Route>
              </Routes>
            ) : (
              <Component {...componentProps} />
            )}
          </Context.Provider>
        </MemoryRouter>
      </Provider>
    </IntlProvider>
  );
};

ComponentWithContext.propTypes = {
  Component: PropTypes.element,
  componentProps: PropTypes.object,
  renderOptions: PropTypes.object,
  Context: PropTypes.object,
  contextValue: PropTypes.object,
};
