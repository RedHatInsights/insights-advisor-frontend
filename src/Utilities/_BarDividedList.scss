  .verticalDivider {
    color: var(--pf-global--BorderColor--100)
  }