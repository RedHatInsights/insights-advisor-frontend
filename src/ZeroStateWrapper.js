import React, { Suspense, useState, useEffect, createContext } from 'react';
import { Bullseye, Spinner } from '@patternfly/react-core';
import AsynComponent from '@redhat-cloud-services/frontend-components/AsyncComponent';
import ErrorState from '@redhat-cloud-services/frontend-components/ErrorState';
import PropTypes from 'prop-types';
import { addNotification } from '@redhat-cloud-services/frontend-components-notifications/redux/actions/notifications';
import { useDispatch } from 'react-redux';
import { useIntl } from 'react-intl';
import messages from './Messages';
import { useFeatureFlag } from './Utilities/Hooks';
import {
  useGetEdgeDevicesQuery,
  useGetConventionalDevicesQuery,
} from './Services/SystemVariety';

export const AccountStatContext = createContext({
  hasConventionalSystems: true,
  hasEdgeDevices: false,
});

export const ZeroStateWrapper = ({ children }) => {
  const intl = useIntl();
  const dispatch = useDispatch();
  const isEdgeParityEnabled = useFeatureFlag('advisor.edge_parity');
  const [hasConventionalSystems, setHasConventionalSystems] = useState(true);
  const [hasEdgeDevices, setHasEdgeDevices] = useState(true);
  const {
    data: edge,
    isSuccess: edgeQuerySuccess,
    isError: edgeError,
    error: edgeErrorMessage,
  } = useGetEdgeDevicesQuery();
  const {
    data: conventional,
    isSuccess: conventionalQuerySuccess,
    isError: conventionalError,
    error: conventErrorMessage,
  } = useGetConventionalDevicesQuery();
  useEffect(() => {
    setHasEdgeDevices(edge?.total > 0);
    setHasConventionalSystems(conventional?.total > 0);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [edgeQuerySuccess, conventionalQuerySuccess]);

  useEffect(() => {
    if (edgeErrorMessage?.status || conventErrorMessage?.status) {
      dispatch(
        addNotification({
          variant: 'danger',
          dismissable: true,
          title: intl.formatMessage(messages.error),
          description:
            `${JSON.stringify(edgeErrorMessage?.data)}` ||
            `${JSON.stringify(conventErrorMessage?.data)}`,
        })
      );
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [edgeError, conventionalError]);

  const hasSystems = isEdgeParityEnabled
    ? hasEdgeDevices || hasConventionalSystems
    : hasConventionalSystems;

  return (
    <Suspense
      fallback={
        <Bullseye>
          <Spinner size="xl" />
        </Bullseye>
      }
    >
      {!hasSystems && conventionalQuerySuccess ? (
        <Suspense
          fallback={
            <Bullseye>
              <Spinner size="xl" />
            </Bullseye>
          }
        >
          <AsynComponent
            appId={'advisor_zero_state'}
            appName="dashboard"
            module="./AppZeroState"
            scope="dashboard"
            ErrorComponent={<ErrorState />}
            app="Advisor"
          />
        </Suspense>
      ) : (
        <AccountStatContext.Provider
          value={{ hasConventionalSystems, hasEdgeDevices, edgeQuerySuccess }}
        >
          {children}
        </AccountStatContext.Provider>
      )}
    </Suspense>
  );
};

ZeroStateWrapper.propTypes = {
  children: PropTypes.element,
  check: PropTypes.boolean,
};
