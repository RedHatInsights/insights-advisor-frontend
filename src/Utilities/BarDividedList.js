import React from 'react';
import './_BarDividedList.scss';

const BarDividedList = (list) => list.map((element, index) => <React.Fragment key={index}>
    {element}{index + 1 !== list.length && <strong className='verticalDivider'>&nbsp;&#124;&nbsp;</strong>}
</React.Fragment>
);

export default BarDividedList;
