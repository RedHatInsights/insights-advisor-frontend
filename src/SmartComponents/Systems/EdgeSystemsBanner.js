import React, { useContext } from 'react';
import { useIntl } from 'react-intl';
import { Alert } from '@patternfly/react-core';
import messages from '../../Messages';
import { AccountStatContext } from '../../ZeroStateWrapper';

const EdgeSystemsBanner = () => {
  const intl = useIntl();
  const { hasEdgeDevices } = useContext(AccountStatContext);

  return !hasEdgeDevices ? null : (
    <Alert
      variant="info"
      isInline
      style={{ marginBottom: '1.5rem' }}
      title={intl.formatMessage(messages.edgeWarning)}
    />
  );
};

export default EdgeSystemsBanner;
