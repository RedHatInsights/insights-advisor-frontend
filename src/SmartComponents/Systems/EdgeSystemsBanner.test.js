import React from 'react';
import '@testing-library/jest-dom';
import { render, screen } from '@testing-library/react';
import EdgeSystemsBanner from './EdgeSystemsBanner';
import { ComponentWithContext } from '../../Utilities/TestingUtilities';
import { AccountStatContext } from '../../ZeroStateWrapper';

describe('EdgeSystemsBanner', () => {
  it('does render with edge hosts present', () => {
    const { asFragment } = render(
      <ComponentWithContext
        Component={EdgeSystemsBanner}
        Context={AccountStatContext}
        contextValue={{ hasEdgeDevices: true, hasConventionalSystems: true }}
      />
    );

    expect(
      screen.getByLabelText('Immutable Systems Warning')
    ).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it('does not render with no edge hosts', () => {
    const { asFragment } = render(
      <ComponentWithContext
        Component={EdgeSystemsBanner}
        Context={AccountStatContext}
        contextValue={{ hasEdgeDevices: false, hasConventionalSystems: true }}
      />
    );

    expect(
      screen.queryByLabelText('Immutable Systems Warning')
    ).not.toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });
});
