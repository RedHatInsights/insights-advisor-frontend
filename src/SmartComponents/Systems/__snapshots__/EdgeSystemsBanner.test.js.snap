// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EdgeSystemsBanner does not render with no edge hosts 1`] = `<DocumentFragment />`;

exports[`EdgeSystemsBanner does render with edge hosts present 1`] = `
<DocumentFragment>
  <div
    aria-label="Immutable Systems Warning"
    class="pf-c-alert pf-m-inline pf-m-info"
    data-ouia-component-id="OUIA-Generated-Alert-info-1"
    data-ouia-component-type="PF4/Alert"
    data-ouia-safe="true"
    style="margin-bottom: 1.5rem;"
  >
    <div
      class="pf-c-alert__icon"
    >
      <svg
        aria-hidden="true"
        fill="currentColor"
        height="1em"
        role="img"
        style="vertical-align: -0.125em;"
        viewBox="0 0 512 512"
        width="1em"
      >
        <path
          d="M256 8C119.043 8 8 119.083 8 256c0 136.997 111.043 248 248 248s248-111.003 248-248C504 119.083 392.957 8 256 8zm0 110c23.196 0 42 18.804 42 42s-18.804 42-42 42-42-18.804-42-42 18.804-42 42-42zm56 254c0 6.627-5.373 12-12 12h-88c-6.627 0-12-5.373-12-12v-24c0-6.627 5.373-12 12-12h12v-64h-12c-6.627 0-12-5.373-12-12v-24c0-6.627 5.373-12 12-12h64c6.627 0 12 5.373 12 12v100h12c6.627 0 12 5.373 12 12v24z"
        />
      </svg>
    </div>
    <h4
      class="pf-c-alert__title"
    >
      <span
        class="pf-u-screen-reader"
      >
        Info alert:
      </span>
      Immutable systems are not shown in this list.
    </h4>
  </div>
</DocumentFragment>
`;
